src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseAsSpanInsteadOfRangeIndexer.Fixer.cs

```diff
@@ -0,0 +1,111 @@
+                UseAsSpanInsteadOfRangeIndexerAnalyzer.ArrayReadOnlyRuleId,
+                UseAsSpanInsteadOfRangeIndexerAnalyzer.ArrayReadWriteRuleId);
+
+        public sealed override FixAllProvider GetFixAllProvider()
+        {
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(context.Span);
+
+            if (node is null)
+            {
+                return;
+            }
+
+            // The rules are mutually exclusive, so there can't be more than one for the same span:
+            var diagnostic = context.Diagnostics.Single();
+            var targetMethod = diagnostic.Properties.GetValueOrDefault(UseAsSpanInsteadOfRangeIndexerAnalyzer.TargetMethodName);
+
+            if (targetMethod == null)
+            {
+                return;
+            }
+
+            context.RegisterCodeFix(
+                new UseAsSpanInsteadOfRangeIndexerCodeAction(
+                    diagnostic.Id,
+                    targetMethod,
+                    context.Document,
+                    node,
+                    this),
+                diagnostic);
+        }
+
+        protected abstract bool TrySplitExpression(
+            SyntaxNode node,
+            out SyntaxNode toReplace,
+            [NotNullWhen(true)] out SyntaxNode? target,
+            [NotNullWhen(true)] out IEnumerable<SyntaxNode>? arguments);
+
+        private class UseAsSpanInsteadOfRangeIndexerCodeAction : CodeAction
+        {
+            private readonly string _targetMethod;
+            private readonly Document _document;
+            private readonly SyntaxNode _rootNode;
+            private readonly UseAsSpanInsteadOfRangeIndexerFixer _fixer;
+
+            public override string Title { get; } = MicrosoftNetCoreAnalyzersResources.UseAsSpanInsteadOfRangeIndexerTitle;
+
+            public override string EquivalenceKey { get; }
+
+            public UseAsSpanInsteadOfRangeIndexerCodeAction(
+                string ruleId,
+                string targetMethod,
+                Document document,
+                SyntaxNode rootNode,
+                UseAsSpanInsteadOfRangeIndexerFixer fixer)
+            {
+                _targetMethod = targetMethod;
+                _document = document;
+                _rootNode = rootNode;
+                _fixer = fixer;
+                EquivalenceKey = ruleId;
+            }
+
+            protected override async Task<Document> GetChangedDocumentAsync(CancellationToken cancellationToken)
+            {
+                var editor = await DocumentEditor.CreateAsync(_document, cancellationToken).ConfigureAwait(false);
+
+                if (_fixer.TrySplitExpression(_rootNode, out var toReplace, out var target, out var arguments))
```

Comment:
Given that the registered code action can bail out, I'd recommend that you check if TrySplitExpression can succeed prior to invoking `RegisterCodeFix`. Otherwise, light bulb will show up, but the preview/application of the fix will have no changes, which might feel like a bug to the customer.